### PR TITLE
refactor: Use intermediate representation of `SELECT` in InfluxQL planner

### LINKED ISSUE #7750: InfluxQL-IOx: Column alias incorrect for binary expression with functions
The column alias for binary expressions is incorrect when specifying a binary expression using functions.

## Steps

1. Execute the following InfluxQL query

    ```sql
    SELECT COUNT(usage) + SUM(usage_idle) FROM cpu
    ```

## What happened?

A column alias that does not match InfluxQL OG:

```text
name: cpu
time                 count_usage_sum_usage_idle
----                 --------------------------
1970-01-01T00:00:00Z
```

## Expected?

The column alias should be `count_sum`:

```text
name: cpu
time                 count_sum
----                 ---------
1970-01-01T00:00:00Z
```

### PR DESCRIPTION
**Pull request by stuartcarnie**
Helps #7739
Closes #7750

This PR involves some refactoring of the InfluxQL planner to use the intermediate representation of an InfluxQL `SELECT` statement. This intermediate representation will continue to evolve over subsequent PRs.

As a bonus, issue #7750 is closed, resolving an inconsistency with column aliases for binary expressions between IOx and InfluxQL OG.

### CODE REVIEW

File: influxdb_iox/tests/query_tests2/cases/in/issue_6112.influxql.expected [hunk 1]
```diff
@@ -1003,11 +1003,11 @@ tags: non_existent=
 +---------------------+-------+--------------------+--------------------+
 -- InfluxQL: SELECT COUNT(f64), COUNT(f64) + COUNT(f64), COUNT(f64) * 3 FROM m0;
 name: m0
-+---------------------+-------+---------------------+-----------+
-| time                | count | count_f64_count_f64 | count_f64 |
-+---------------------+-------+---------------------+-----------+
-| 1970-01-01T00:00:00 | 7     | 14                  | 21        |
-+---------------------+-------+---------------------+-----------+
++---------------------+-------+-------------+---------+
+| time                | count | count_count | count_1 |
```
Comment (stuartcarnie) [inline influxdb_iox/tests/query_tests2/cases/in/issue_6112.influxql.expected:1007]: A fix for #7750

File: influxdb_iox/tests/query_tests2/cases/in/issue_6112.influxql.expected
Comment (stuartcarnie) [inline influxdb_iox/tests/query_tests2/cases/in/issue_6112.influxql.expected:1]: The changes to this file are a result of a fix for #7750

File: iox_query_influxql/src/plan/field.rs [hunk 2]
```diff
@@ -66,7 +66,7 @@ impl<'a> Visitor for BinaryExprNameVisitor<'a> {
 
     fn pre_visit_call(self, n: &Call) -> Result<Recursion<Self>, Self::Error> {
         self.0.push(n.name.clone());
-        Ok(Recursion::Continue(self))
+        Ok(Recursion::Stop(self))
```
Comment (stuartcarnie) [inline iox_query_influxql/src/plan/field.rs:69]: The fix for #7750

File: iox_query_influxql/src/plan/planner.rs
Comment (stuartcarnie) [inline iox_query_influxql/src/plan/planner.rs:1]: The changes to the tests were also a result of the fix for #7750

File: iox_query_influxql/src/plan/rewriter.rs
Comment (stuartcarnie) [inline iox_query_influxql/src/plan/rewriter.rs:1]: I reorganised the `rewrite_statement` tests into separate scenarios, besides the fix for #7750, the output remained consistent before this PR as did the end-to-end tests.

File: iox_query_influxql/src/plan/rewriter.rs [hunk 3]
```diff
@@ -58,6 +57,24 @@ pub(super) fn map_select(s: &dyn SchemaProvider, stmt: &SelectStatement) -> Resu
...
+///
+/// # `SLIMIT` and `SOFFSET`
+///
+/// * `SLIMIT` and `SOFFSET` don't work as expected per issue [#7571]
+/// * This issue [is noted](https://docs.influxdata.com/influxdb/v1.8/query_language/explore-data/#the-slimit-clause) in our official documentation
+///
+/// [#7571]: https://github.com/influxdata/influxdb/issues/7571
+fn check_features(stmt: &SelectStatement) -> Result<()> {
+    if stmt.series_limit.is_some() || stmt.series_offset.is_some() {
+        return error::not_implemented("SLIMIT or SOFFSET");
```
Comment (mhilton) [inline iox_query_influxql/src/plan/rewriter.rs:72]: We could easily have a separate error for `SLIMIT`and `SOFFSET` is there a reason not to? I don't think this should be changed, I'm just intrigued.

**Review by mhilton — APPROVED** [review]

File: iox_query_influxql/src/plan/rewriter.rs (hunk 3 above)
Comment (stuartcarnie) [reply iox_query_influxql/src/plan/rewriter.rs:72]: For reference, I transferred the message from here:

https://github.com/influxdata/influxdb_iox/blob/af726e092ffee2213081b26cd2b3d5533b944c28/iox_query_influxql/src/plan/planner.rs#L881-L885

My reasoning is that the two clauses are related, if we ever do support one, we'll support the other and a single message lets them know that both are not supported so don't try the other if you've tried one 😆

Comment (stuartcarnie) [reply iox_query_influxql/src/plan/rewriter.rs:72]: For reference, the most compete version for all variants would be:


```rust
match (stmt.series_limit, stmt.series_offset) {
  (Some(_), Some(_)) => return error::not_implemented("SLIMIT or SOFFSET"),
  (Some(_), None) => return error::not_implemented("SLIMIT"),
  (None, Some(_)) => return error::not_implemented("SOFFSET"),
  (None, None) => {}
}
```

Arguably, the specific errors don't measurably improve the user experience to warrant the extra code.

**Pull request merged**